parse_cmd_test.go

```diff
@@ -1,151 +1,134 @@
-	{"unskip_wrong_usage", "unskip", "help", nil, true},
-	{"version_extra_args_ok", "version info", "version", nil, false},
+var rejectedCommands = []string{
+	"",
+
+	// Funnily enough, these *do* give you what you want!
+	"help me",
+	"halp",
+	"schedule help",
+
+	// Unexpected arguments
+	"status me",
+	"cookie me",
+
+	// Did they really want `schedule`?
+	"subscribe tue",
+	"unsubscribe thu",
+
+	// (Un)skipping requires an argument.
+	"skip",
+	"unskip",
+
+	// TODO(#49): Allow (un)skipping days other than tomorrow
+	"skip friday",
+	"unskip next",
+
+	// This is not the way to delete reviews you don't like 😛
+	"get-reviews -1",
+	"get-reviews -10",
+
+	// Unknown commands
+	"scheduleing monday",
+	"schedul monday",
+	"mooh",
 }
 
-func TestParseCmdWithArgs(t *testing.T) {
-	for _, tt := range tableWithArgs {
-		t.Run(tt.testName, func(t *testing.T) {
-			gotCmd, gotArgs, gotErr := parseCmd(tt.inputStr)
-			if gotCmd != tt.wantedCmd || len(gotArgs) != len(tt.wantedArgs) {
-				t.Errorf("got %v, %v, wanted %v, %v\n", gotCmd, gotArgs, tt.wantedCmd, tt.wantedArgs)
+func Test_parseCmd(t *testing.T) {
+	for input, expected := range acceptedCommands {
+		t.Run(input, func(t *testing.T) {
+			cmd, args, err := parseCmd(input)
+			if err != nil {
+				t.Fatalf("unexpected error: %#+v", err)
 			}
 
-			switch gotCmd {
-			case "schedule":
-				for i, arg := range gotArgs {
-					if _, ok := daysList[arg]; !ok {
-						t.Errorf("Wrong argument %v for command %v\n", gotArgs[i], gotCmd)
-					}
-				}
-			case "skip":
-				if gotArgs[0] != "tomorrow" {
-					t.Errorf("Wrong argument %v for command %v\n", gotArgs[0], gotCmd)
-				}
-			case "unskip":
-				if gotArgs[0] != "tomorrow" {
-					t.Errorf("Wrong argument %v for command %v\n", gotArgs[0], gotCmd)
-				}
-			case "version":
-				// Always fine as long as the wantedCmd check passes.
-			default:
-				if gotCmd != "help" {
-					t.Errorf("unknown command %v\n", gotCmd)
-				}
-			}
+			assertEqual(t, cmd, expected.Cmd)
+			assertEqual(t, args, expected.Args)
+		})
+	}
 
-			_, ok := gotErr.(*parsingErr)
+	for _, input := range rejectedCommands {
+		t.Run(input, func(t *testing.T) {
+			cmd, args, err := parseCmd(input)
 
-			if tt.expectErr && !ok {
-				t.Errorf("Expected parsingErr but didn't get one\n")
-			} else if !tt.expectErr && ok {
-				t.Errorf("Got unexpected parsingError\n")
-			}
+			_, _ = assertErrorAs[*parsingErr](t, err)
+
+			assertEqual(t, cmd, "help")
+			assertEqual(t, args, nil)
 		})
 	}
 }
 
-var tableMisc = []struct {
-	testName   string
-	inputStr   string
-	wantedCmd  string
-	wantedArgs []string
-	expectErr  bool
-}{
-	{"command_is_superstring", "scheduleing monday", "help", nil, true},
-	{"command_is_substring", "schedul monday", "help", nil, true},
-	{"command_is_undefined", "mooh", "help", nil, true},
-	{"command_is_capitalized", "Help", "help", nil, false},
-}
+func assertEqual[T any](t *testing.T, a, b T) {
```

Comment:
There's a recommendation in the [Golang style guide](https://google.github.io/styleguide/go/decisions#equality-comparison-and-diffs) to use [`cmp`](https://pkg.go.dev/github.com/google/go-cmp/cmp) for this kind of comparison: "a more powerful and safer alternative to [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual) for comparing whether two values are semantically equal."

That said, "bring in another dependency" eh?

Reply:
Yeah, I've been trying _really_ hard not to add my usual testing dependencies. This is basically a copy of testify's [assert.Equal](https://pkg.go.dev/github.com/stretchr/testify/assert), which is what I've been used to using for a while.

I haven't imported testify itself mostly because we like this being a low-dependency project, but also because my muscle memory for this is pre-generics, and I haven't looked for good new-style helpers yet.

I think `cmp`'s diff output is a big enough improvement to error reporting that it's worth adding. I'll do it in a separate sweep to clear out all the `assertEqual`s I've added along the way.